cgsmiles/graph_layout.py

```diff
@@ -0,0 +1,222 @@
+    energy = np.sum(np.power(dist_matrix, -8))
+    return energy
+
+
+def optimize_geometry_2D(pos, interactions, inter_methods):
+    """
+    Given a position array of shape (n, 2) and a dict of
+    interactions, optimize the positions to minimize the
+    energy.
+
+    Parameters
+    ----------
+    pos: np.ndarray((n, 2))
+        the inital positions
+    interactions: dict[str, list(tuple)]
+        a dict of interactions with key type and list of
+        interaction tuples
+
+    Returns
+    -------
+    np.ndarray((n, 2)), float
+        the optimized positions and final energy
+    """
+    n_atoms = len(pos)
+    atom_to_idx = OrderedDict(zip(list(pos.keys()), range(0, n_atoms)))
+    positions = np.ravel(np.array(list(pos.values())))
+
+    def target_function(positions):
+        energy = 0
+        positions = positions.reshape((-1, 2))
+        for inter_type, inters in interactions.items():
+            for inter in inters:
+                atom_pos = [positions[atom_to_idx[ndx]] for ndx in inter[:-1]]
+                energy += inter_methods[inter_type](atom_pos, inter[-1])
+        energy += nonbonded_potential(positions)
+        return energy
+
+    opt_result = scipy.optimize.minimize(target_function,
+                                         positions,
+                                         method='L-BFGS-B',
+                                         options={'ftol':0.01, 'maxiter':500})
+
+    positions_tmp = opt_result['x'].reshape((-1, 2))
+    positions = {}
+    for node_key, idx in atom_to_idx.items():
+        positions[node_key] = positions_tmp[idx]
+    return positions, opt_result['fun']
+
+
+class MoleculeLayouter2D:
+    """
+    Generate 2D coordinate for drawing a molecule.
+    """
+    def __init__(self, graph, max_iter=5, target_energy=1000):
+        self.inter_methods = {'bonds': bond_potential, 'angles': angle_potential}
+        self.graph = graph
+        # some hyperparameters
+        self.max_iter = max_iter
+        self.target_energy = target_energy
+
+    def force_minimize(self, pos):
+        bonds = assign_bonds(self.graph, pos)
+        angles = assign_angles(self.graph)
+        interactions = {'bonds': bonds, 'angles': angles}
+        pos, energy = optimize_geometry_2D(pos, interactions, self.inter_methods)
+        return pos, energy
+
+    def vsepr_layout(self):
+        dist = dict(nx.shortest_path_length(self.graph, weight=None))
+        for source, dest_d in dist.items():
+            for dest, d in dest_d.items():
+                if d % 2 == 1:  # odd
+                    # sqrt(1 + (n*x)**2 + n*x**2)
+                    n = (d + 1)/2
+                    dist[source][dest] = math.sqrt(3*n**2 - 3*n + 1)
+                else:  # even
+                    dist[source][dest] =  d / 2 * math.sqrt(3)
+        # do we need this try except????
+        try:
+            p0 = nx.fruchterman_reingold_layout(self.graph)
+        except nx.NetworkXException:
+            p0 = None
```

Comment:
I'm not sure doing the spring layout to initialize the kamada_kawai to initialize the optimization is super useful. I'd get rid of the spring

Reply:
it actually helps for large molecules